### PR TITLE
New connected components implementation

### PR DESCRIPTION
**Pull request by mengxr**
This PR replaces the GraphX implementation of connected components by a DataFrame-based implementation. It also switched the algorithm to the one proposed in "Connected Components in MapReduce and Beyond". The new implementation handles skewness by separating the joins into broadcast join of vertices of high degrees and normal join of the rest.

The implementation was tested on a graph w/ ~2 billion edges, but without long ID conversion.

Note that it uses `indexedVertices` and `indexedEdges` to assign long IDs. This might have performance issues if the input graph is skewed. We can do a follow-up PR to address it.

The unit tests are updated to check the broadcast join path. The coverage is apparently not ideal at this time.

cc: @jkbradley

### CODE REVIEW

**Comment by mengxr** [comment]
This PR depends on #120. Otherwise, it will fail in 1.4 and 1.5 builds.

**Comment by nchammas** [comment]
Couple of quick questions:

1. Will PySpark automatically benefit from this improved implementation, or is separate work required for that?
2. Is a notable performance difference expected with this new implementation?

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 1]
```diff
@@ -17,38 +17,262 @@
...
+    var prevSum = Long.MaxValue
+    while (!converged) {
+      val toBeCleaned = mutable.ArrayBuffer.empty[DataFrame]
+      toBeCleaned += ee
+      // compute min neighbors including self
+      val minNbrs1 = minNbrs(ee)
+        .withColumn(MIN_NBR, minValue(col(SRC), col(MIN_NBR)).as(MIN_NBR))
+        .persist(StorageLevel.MEMORY_AND_DISK)
+      toBeCleaned += minNbrs1
+      // connect all strictly larger neighbors to the min neighbor (including self)
```
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: Note that this is the large-star op

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 2]
```diff
@@ -17,38 +17,262 @@
...
+      ee = skewedJoin(ee, minNbrs1, broadcastThreshold, logPrefix)
+        .select(col(DST).as(SRC), col(MIN_NBR).as(DST)) // src > dst
+        .distinct()
+        .persist(StorageLevel.MEMORY_AND_DISK)
+      toBeCleaned += ee
+      // compute min neighbors
+      val minNbrs2 = minNbrs(ee)
+        .persist(StorageLevel.MEMORY_AND_DISK)
+      toBeCleaned += minNbrs2
+      // connect all smaller neighbors to the min neighbor
```
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: Note that this is the small-star op

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 3]
```diff
@@ -17,38 +17,262 @@
...
+    val vv = g.vertices
+    var ee = g.edges.persist(StorageLevel.MEMORY_AND_DISK)
+    val numEdges = ee.count()
+    logger.info(s"$logPrefix Found $numEdges edges after preparation.")
+
+    var converged = false
+    var k = 0
+    var prevSum = Long.MaxValue
+    while (!converged) {
+      val toBeCleaned = mutable.ArrayBuffer.empty[DataFrame]
```
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: toBeCleaned -> toBeUnpersisted

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 4]
```diff
@@ -17,38 +17,262 @@
...
+        .distinct()
+      // TODO: remove this after DataFrame.checkpoint is implemented
+      val out = s"$checkpointDir/$k"
+      ee.write.parquet(out)
+      if (k > 0) {
+        FileSystem.get(sc.hadoopConfiguration)
+          .delete(new Path(s"$checkpointDir/${k - 1}"), true)
+      }
+      toBeCleaned.foreach(df => df.unpersist(true))
+      System.gc() // hint Spark to clean shuffle directories
```
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: Map GC hint across workers too?

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 5]
```diff
@@ -17,38 +17,262 @@
...
+ *
  * The resulting DataFrame contains all the vertex information and one additional column:
  *  - component (`LongType`): unique ID for this component
  *
  * The resulting edges DataFrame is the same as the original edges DataFrame.
+ *
+ * @see https://mmds-data.org/presentations/2014/vassilvitskii_mmds14.pdf
  */
-class ConnectedComponents private[graphframes] (private val graph: GraphFrame) extends Arguments {
+class ConnectedComponents private[graphframes] (
```
Comment (mengxr) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: done

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 6]
```diff
@@ -17,38 +17,262 @@
...
+      .select(SRC, MIN_NBR)
+      .as[(Long, Long)]
+      .collect()
+      .toMap // TODO: use OpenHashMap
+    if (hubs.isEmpty) {
+      return edges.join(minNbrs, SRC)
+    } else {
+      logger.debug(s"$logPrefix Number of skewed keys: ${hubs.size}.")
+    }
+    val isHub = udf { id: Long =>
```
Comment (mengxr) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: I don't quite understand this comment.

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 7]
```diff
@@ -17,38 +17,262 @@
...
+    hashJoined.unionAll(broadcastJoined)
+  }
+
+  /**
+   * Runs connected components with default parameters.
+   */
   def run(graph: GraphFrame): DataFrame = {
-    val gx = graphxlib.ConnectedComponents.run(graph.cachedTopologyGraphX)
-    GraphXConversions.fromGraphX(graph, gx, vertexNames = Seq(COMPONENT_ID)).vertices
+    new ConnectedComponents(graph).run()
```
Comment (mengxr) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: Good point. Shall I make another PR for it?

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 8]
```diff
@@ -17,38 +17,262 @@
...
   }
 
-  private val COMPONENT_ID = "component"
+  private def run(graph: GraphFrame, broadcastThreshold: Int): DataFrame = {
+    val runId = UUID.randomUUID().toString.takeRight(8)
+    val logPrefix = s"[CC $runId]"
+    logger.info(s"$logPrefix Start connected components with run ID $runId.")
```
Comment (mengxr) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: `logger.info` uses macro so it doesn't unnecessarily materialize the message. I'm not sure whether `logInfo` behaves the same.

**Comment by mengxr** [comment]
@nchammas The python interface should use the new implementation automatically, though I haven't update it to take `broadcastThreshold` as an arg.

The new implementation pays more attention to scalability issues we observed on huge graphs. For smaller graphs, it could be slower due to skewed joins and checkpointing. We are working on several possible optimizations, including:

1) DataFrame checkpointing
2) support checkpoint interval
3) better/reliable long ID assignment

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 9]
```diff
@@ -17,38 +17,275 @@
...
  * Connected components algorithm.
  *
  * Computes the connected component membership of each vertex and returns a DataFrame of vertex
  * information with each vertex assigned a component ID.
  *
+ * The computation is done using alternating large star and small star iterations proposed in
+ * "Connected Components in MapReduce and Beyond" with skewed join optimization.
+ * Intermediate data is checkpointed under [[org.apache.spark.SparkContext.getCheckpointDir]]
+ * with prefix "connected-components".
+ * If the checkpoint directory is not set, this throws nn [[IOException]].
```
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: nn -> an

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala (hunk 6 above)
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: (discussed offline, here for the record:)  We could leave the hubs in a DataFrame instead of collecting them as a map.  Then the broadcast join could be done using a broadcast hint.  To be done eventually once we believe it is more efficient.

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala (hunk 8 above)
Comment (jkbradley) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: logInfo calls logger.info, just as in Spark.  not important though

**Comment by jkbradley** [comment]
Thanks @mengxr !  Merging now

**Comment by jkbradley** [comment]
Linking with issue: https://github.com/graphframes/graphframes/issues/117
